crates/subspace-farmer/src/farmer_piece_getter.rs

```diff
@@ -438,6 +439,40 @@ impl<FarmIndex, CacheIndex, PV, NC> Clone for WeakFarmerPieceGetter<FarmIndex, C
     }
 }
 
+/// This wrapper allows us to return the stream, which in turn depends on `piece_getter` that was
+/// previously on the stack of the inner function. What this wrapper does is create a
+/// self-referential data structure, so we can move both together, while still implementing `Stream`
+/// trait as necessary.
+#[ouroboros::self_referencing]
+struct StreamWithPieceGetter<FarmIndex, CacheIndex, PV, NC>
```

Comment:
Is there anything more than a convenient approach here? We introduce a new third-party crate with insufficient features that doesn't seem to be critical. You generally oppose these kinds of solutions.

Reply:
The only workaround I could come up with would be to not do stream request and instead do a sequence of individual requests like fallback implementation in tests is doing. And it seemed not great to me despite I don't think we actually use streaming version of weak piece getter, but for completeness of the implementation it had to be done.

I agree it is suboptimal and I was hoping to convince the compiler to "extend" the lifetime of `piece_getter` with various tricks, but it didn't believe me. I totally expect this to not be necessary in the future, but for now it is what it is.

I can come up with other ugly workarounds like creating a version of `get_pieces` method that takes `self` instead of `&self`, but it is semantically wrong and I like it even less. This fix while annoying, is localized and limited.